signing_service/src/signing_service.py

```diff
@@ -0,0 +1,139 @@
+        self.port = port
+        self.reconnect_delay = reconnect_delay
+        self.current_reconnect_delay = 0
+        self.socket: socket.socket = None  # type: ignore
+        self.is_connection_established = False
+
+        # Handle shutdown signal.
+        signal.signal(signal.SIGTERM, self._terminate_connection_and_exit)  # type: ignore
+
+        self._run()
+
+    def _run(self) -> None:
+        """
+        Handles main connection loop and its exceptions.
+
+        If connection is interrupted due to a failure, it should try to reconnect.
+        If connection is closed by other side, exit gracefully.
+        If a shutdown signal or KeyboardInterrupt is caught, exit gracefully.
+        If there was an unrecognized exception, it logs it and report to Sentry, then reraise and crash.
+        """
+        while True:
+            try:
+                if self.is_connection_established is False:
+                    self._connect()
+            except socket.error as exception:
+                logger.error(f'Socket error occurred: {exception}')
+
+                # When the other side closes the connection, return without errors
+                if isinstance(exception.args, tuple) and exception.args[0] == socket.errno.EPIPE:  # type: ignore
+                    self._terminate_connection_and_exit()
+
+                # Increase delay and reconnect if the connection is interrupted due to a failure.
+                self._increase_delay()
+                self.is_connection_established = False
+            except KeyboardInterrupt:
+                # Handle keyboard interrupt.
+                self._terminate_connection_and_exit()
+            except Exception as exception:
+                # Try to close opened socket.
+                self.socket.close()
+                # If there was an unrecognized exception, log it and report to Sentry.
+                logger.error(f'Unrecognized exception occurred: {exception}')
+                crash_logger.error(str(exception))
+                raise
+
+    def _connect(self) -> None:
+        """ Creates socket and connects to given HOST and PORT. """
+        self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+
+        if self.current_reconnect_delay > 0:
+            logger.warning(f'Waiting {self.current_reconnect_delay} before connecting.')
+            sleep(self.current_reconnect_delay)
+
+        logger.warning(f'Connecting to {self.host}:{self.port}.')
+        self.socket.connect((self.host, self.port))
+        logger.warning(f'Connection established.')
+
+        # Reset delay on successful connection and set flag that connection is established.
+        self.current_reconnect_delay = 0
+        self.is_connection_established = True
+        self._handle_connection()
+
+    def _handle_connection(self) -> None:
+        """ Inner loop that handles data exchange over socket. """
+        while True:
+            data = self.socket.recv(1024)
```

Comment:
you could use `asyncio` coroutines here

Reply:
Yes, will be added.

Reply:
@dybi @cameel 
I have added it in separate commit https://github.com/golemfactory/concent/pull/634/commits/09b3193ecccff5257c1e4fd04b4e6274489f3b7d so you can either review it as a whole or only those changes. And I can drop it if we decide it to be not needed after all.

Reply:
@dybi @cameel Please remind me what we finally agreed on ? Because if it is up to me I would like not to add it as a part of this pull request, because
- it will not have big impact on how this client actually works looking from outside of it,
- it will be difficult to test it,
- it will require rewriting most of the current code.

So if we really want to do this, I suggest to do this in separate issue. If it is optional, I would skip it for now.

Reply:
Separate issue is fine. It is optional so if you don't want it, you can skip it altogether, though I would recommend doing it just to practice using asyncio a bit.

Reply:
:+1: done https://github.com/golemfactory/concent/issues/647